### PR TITLE
Default Filament Runout Sensor enabled state

### PR DESCRIPTION
**Pull request by CRCinAU**
### Requirements

When a binary is compiled without `FILAMENT_RUNOUT_SENSOR`, the structure in the EEPROM defaults to a value of 'true' - which maps back into `runout.enabled`. This means that if a firmware is loaded later that has `FILAMENT_RUNOUT_SENSOR` enabled as a build time option, the sensor is automatically enabled just because the build option was present.

This patch set the default to disabled when built without `FILAMENT_RUNOUT_SENSOR`, and upon a reset with `M502`, resets the value to false.

This allows firmware to be shipped with support for a filament runout sensor - but disabled by default - meaning an end user need only plug the sensor in, and then enable with `M412 S1`.

This is a more sane default than the inverse, which requires the user to turn the option off manually - or if a firmware is shipped, it needs to be shipped without `FILAMENT_RUNOUT_SENSOR` enabled and a second binary provided to enable the sensor.

From my understanding of how the defaults / settings part works, I believe these are the only two parts that require modification to achieve this goal.

### CODE REVIEW

**Comment by thinkyhead** [comment]
The problem I have found is that people (well, me…) routinely forget(s) to turn on the filament runout sensor after resetting the EEPROM, and I also don't really want to put `M412` into my starting G-code. So the most convenient default in that light is to have the sensor enabled by default on an EEPROM reset. The default `enabled` value was `true` before EEPROM storage was added in #13876, and it is still the default if you disable EEPROM storage today.

Whenever there is a controversy over the "best" behavior, the answer is usually to add a new configuration option. But I hesitate to add new options, especially if they alter low-level behaviors, if only because it erodes the consistency of behavior from one Marlin-based printer to another. But maybe this is a lesser concern with things like filament sensors….

**Comment by CRCinAU** [comment]
I fully understand the reasons why it was enabled by default - however I feel that it hinders the feature being built in by default - as every user who doesn't have a filament sensor (which is most), have to go and turn the feature off if they don't have it.

My working case is the new Creality Ender 3 V2 - it has a filament runout port - and only a few people have added runout sensors - which means everyone who flashes a firmware with support for that port has to then turn the sensor that they don't have off to keep things sane.

Conversely, if a user wants to add a filament sensor later on, they can then just plug the sensor in, turn the feature on with `S412 S1`, and hey presto - it works - no firmware changes required.

In this real world example, I feel that the number of users without the sensor would greatly outnumber those that have the sensor.

Documentation for said runout sensors can easily state something like "When you plug the sensor in, make sure you turn it on via .........." instead of board manufacturers / firmware builder having to say "If you don't have a sensor, make sure you disable that functionality via S412 S0".

**Comment by CRCinAU** [comment]
tl;dr - If you add some hardware, it's a good expectation to turn the feature on too. Less acceptable to turn off a hardware option that you don't have or your printer won't work properly.

**Comment by thinkyhead** [comment]
Not to be flip, but if a vendor wants to change the `true` to `false` in their shipping Marlin build for the reason you cite — well, they do have the source code, and vendors are known to do all kinds of mods that are even more elaborate.

The conclusion seems to be that some will prefer `true` and others will prefer `false` so it only makes sense to add a new configuration option specifying the preferred fallback state. It is currently `true` so that should remain the default here and in the example configurations.

**Comment by CRCinAU** [comment]
Fully agreed - however if they flash a firmware from a different party - or even build their own, it will flip the default back to ON again in the EEPROM.

ie: it leads to a fragmented and inconsistent end user experience.

EDIT: Actually, backing up a little - it would probably improve the situation even if the default for `runout.enabled` is `false` in the eeprom structure ONLY when `FILAMENT_RUNOUT_SENSOR` is not defined.

File: Marlin/src/module/settings.cpp [hunk 1]
```diff
@@ -2469,7 +2469,7 @@ void MarlinSettings::reset() {
   //
 
   #if HAS_FILAMENT_SENSOR
-    runout.enabled = true;
+    runout.enabled = false;
```
Comment (tpruvot) [inline Marlin/src/module/settings.cpp]: please do not disable it.... by default

Comment (CRCinAU) [reply Marlin/src/module/settings.cpp]: I'm curious about the logic on why? There are many board that ship with a filament sensor port - and the majority never get connected to a filament sensor and remain unpopulated. Why would it not be sane to have a default of off - and not require every user with a board that features a filament sensor which remains unused to turn the feature off or they won't be able to print?

This leaves the upgrade path for an end user to be able to upgrade to using a filament sensor by plugging it in and turning it on.

Personal preferences aside, I can't see the reasoning behind leaving it as a 'broken by default' behaviour for printing.

**Comment by Swishknee** [comment]
So my take is having it on by default is incorrect. For those who have installed the sensor, they are aware of it and its needs. Or at the very least the need to research the needs. Those who pull firmware configs, may not be, and will subsequently be faced with a perceived bug or issue. We don’t have ZProbe settings on by default, nor is sensor-less homing. So why have something on, that in a default case should not be?

File: Marlin/src/module/settings.cpp (hunk 1 above)
Comment (tpruvot) [reply Marlin/src/module/settings.cpp]: In this case, maybe a new config define to set it on by default is required... Most "ready to use"printers have it these days... sure DIY ones are different (standalone motherboards)

Comment (Swishknee) [reply Marlin/src/module/settings.cpp]: Most “ready to use printers” are not hacked by industry users (Ultimaker, Lulzbot, and Prusa to an extent). I have used all 3 professionally and at that level you don’t hack them, you use them for the intended purpose and buy a new printer if you need other capabilities. 

You’re argument for not changing it is to take an open source firmware and tailor it to private industry. Who have their own engineering teams to build, test, and debug their releases. If you start talking about ready to use printers in terms of Creality, Anet, etc. Then I would say I have yet to see someone who bought one and kept it stock. But at that point, without solid data, it is impossible to say what the huge community actually does or dosn’t do. But why not leave it up to them?

Also. none of those printers in the latter group come with filament runout sensors as of yet. So an argument to tailor to hackable printers and the community is appropriate, but in that case it would be default off. 

For those wise enough to put a sensor on, they would be aware of the need to support it. Hiding a setting from users who would not be looking for it is as one person has said on a facebook group “is a hostile default”.

**Comment by sjasonsmith** [comment]
After discussing it with a few people on Discord, I think a new option for the default should be added. I dislike any setting which is stored to EEPROM but cannot be configured through the header file. I update firmware constantly, and find it very annoying to have to use gcode to restore EEPROM state every time I reset it.

**Comment by CRCinAU** [comment]
> After discussing it with a few people on Discord, I think a new option for the default should be added. I dislike any setting which is stored to EEPROM but cannot be configured through the header file. I update firmware constantly, and find it very annoying to have to use gcode to restore EEPROM state every time I reset it.

This is ~~exasperated~~ exacerbated by the fact that even if a firmware is built WITHOUT support for a filament sensor, the EEPROM value is still set to ENABLED - therefore any future flashing of firmware that *does* support a sensor will have it enabled as soon as the flash is complete.

This is also a stealth operation to pop up unexpected and prevent the user from printing.

**Comment by tpruvot** [comment]
i suggest do to like the PLR option :
Marlin/Configuration_adv.h:1157:    **#define PLR_ENABLED_DEFAULT   false** // Power Loss Recovery enabled by default.

so **FIL_RUNOUT_DEFAULT** maybe

**Comment by thisiskeithb** [comment]
> i suggest do to like the PLR option

Hey! I suggested that :p

**Comment by CRCinAU** [comment]
I would be agreeable to this ONLY if the setting in the EEPROM structure is set as FALSE when the firmware compiled does not have `FILAMENT_RUNOUT_SENSOR` defined.

**Comment by sjasonsmith** [comment]
> I would be agreeable to this ONLY if the setting in the EEPROM structure is set as FALSE when the firmware compiled does not have `FILAMENT_RUNOUT_SENSOR` defined.

1. That sounds fine to me, that EEPROM value means nothing without the feature enabled.

2. I think you care way too much about  the scenario where somebody writes new firmware and it already has a value stored. Unless you've recently upgraded firmware, you will almost always have to reset EEPROM after updating.

**Comment by CRCinAU** [comment]
>     2\. I think you care way too much about  the scenario where somebody writes new firmware and it already has a value stored. Unless you've recently upgraded firmware, you will almost always have to reset EEPROM after updating.

Yes and no - you only are forced to do a reset if the EEPROM schema version changes... Without that part, it will just load settings from a prior firmware...

**Comment by sjasonsmith** [comment]
> Yes and no - you only are forced to do a reset if the EEPROM schema version changes... Without that part, it will just load settings from a prior firmware...

For somebody who occasionally updates their firmware the layout will almost always change. If somebody updates their firmware constantly (like me) they are hopefully sophisticated enough to be able to deal with it.

**Comment by sjasonsmith** [comment]
@CRCinAU are you able to update your PR to turn it into a configuration option?

**Comment by CRCinAU** [comment]
I'll have to do some digging to figure out how.... I'm pretty sure I have an idea - but I'll give it a go on Monday....

Do I update the default Configuration.h etc too? Who does the mass update in the Configurations repo?

File: Marlin/src/module/settings.cpp (hunk 1 above)
Comment (thinkyhead) [reply Marlin/src/module/settings.cpp]: I concur. It is not my preference either.

**Comment by thinkyhead** [comment]
> Do I update the default Configuration.h etc too? Who does the mass update in the Configurations repo?

I will take care of that next time I synchronize, which will be later today. It's a partially-automated operation, but I have to manually do search-and-replace to apply the changes to the bulk of the examples.
